components/cdc/src/channel.rs

```diff
@@ -19,17 +30,64 @@ const CDC_MSG_MAX_BATCH_SIZE: usize = 128;
 // 2 = (CDC_MSG_MAX_BATCH_SIZE * 1KB / service::CDC_MAX_RESP_SIZE).ceil() + 1 /* reserve for ResolvedTs */;
 pub const CDC_EVENT_MAX_BATCH_SIZE: usize = 2;
 
-pub fn canal(buffer: usize) -> (Sink, Drain) {
+#[derive(Clone)]
+pub struct MemoryQuota {
+    capacity: usize,
+    in_use: Arc<AtomicUsize>,
+}
+
+impl MemoryQuota {
+    pub fn new(capacity: usize) -> MemoryQuota {
+        MemoryQuota {
+            capacity,
+            in_use: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+    pub fn in_use(&self) -> usize {
+        self.in_use.load(Ordering::Relaxed)
+    }
+    pub fn cap(&self) -> usize {
+        self.capacity
+    }
+    fn alloc(&self, bytes: usize) -> bool {
+        let mut total_bytes = self.in_use.load(Ordering::Relaxed);
+        loop {
+            if total_bytes + bytes > self.capacity {
+                return false;
+            }
```

Comment:
In theory this check could have false positive, when `total_bytes + bytes > self.capacity` in L55, but the `total_bytes` has been decreased actually.